extension/apikeyauthextension/authenticator.go

```diff
@@ -0,0 +1,283 @@
+	//     Authorization: ApiKey base64(ID:APIKey)
+	//
+	encoded, found := strings.CutPrefix(orig, "ApiKey ")
+	if !found {
+		return "", "", errAuthorizationHeaderWrongScheme
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(encoded)
+	if err != nil {
+		return "", "", err
+	}
+	id, _, ok := strings.Cut(string(decoded), ":")
+	if !ok {
+		return "", "", errAuthorizationHeaderInvalid
+	}
+
+	return orig, id, nil
+}
+
+func getHeader(headers map[string][]string, titlecase, lowercase string) (string, bool) {
+	values, ok := headers[titlecase]
+	if !ok {
+		values, ok = headers[lowercase]
+		if !ok {
+			for k, v := range headers {
+				if strings.EqualFold(k, lowercase) {
+					values = v
+					break
+				}
+			}
+		}
+	}
+	if len(values) != 0 {
+		return values[0], true
+	}
+	return "", false
+}
+
+// hasPrivileges checks if the API Key is valid and has the required privileges.
+func (a *authenticator) hasPrivileges(ctx context.Context, authHeaderValue string) (bool, string, error) {
+	applications := make([]types.ApplicationPrivilegesCheck, len(a.config.ApplicationPrivileges))
+	for i, app := range a.config.ApplicationPrivileges {
+		applications[i] = types.ApplicationPrivilegesCheck{
+			Application: app.Application,
+			Privileges:  app.Privileges,
+			Resources:   app.Resources,
+		}
+	}
+	req := a.esClient.Security.HasPrivileges()
+	req.Header(authorizationHeader, authHeaderValue)
+	req.Request(&hasprivileges.Request{Application: applications})
+	resp, err := req.Do(ctx)
+	if err != nil {
+		return false, "", err
+	}
+	return resp.HasAllRequested, resp.Username, nil
+}
+
+// getCacheKey computes a cache key for the given API Key ID and headers.
+func (a *authenticator) getCacheKey(id string, headers map[string][]string) (string, error) {
+	key := id
+	for _, header := range a.config.Cache.KeyHeaders {
+		value, ok := getHeader(headers, header, strings.ToLower(header))
+		if !ok {
+			return "", fmt.Errorf("error computing cache key: missing header %q", header)
+		}
+		key += " " + value
+	}
+	return key, nil
+}
+
+// Authenticate validates an ApiKey scheme Authorization header,
+// passing it to Elasticsearch for checking privileges.
+func (a *authenticator) Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error) {
+	authHeaderValue, id, err := a.parseAuthorizationHeader(headers)
+	if err != nil {
+		return ctx, err
+	}
+
+	cacheKey, err := a.getCacheKey(id, headers)
+	if err != nil {
+		return ctx, err
+	}
+
+	derivedKey := pbkdf2.Key(
+		[]byte(authHeaderValue),
+		a.salt[:],
+		a.config.Cache.PBKDF2Iterations,
+		32, // key length
+		sha512.New,
+	)
+	if err != nil {
+		return ctx, err
+	}
+
+	if cacheEntry, ok := a.cache.Get(cacheKey); ok {
+		if subtle.ConstantTimeCompare(cacheEntry.key, derivedKey) == 0 {
+			// Client has specified an API Key with a colliding ID,
+			// but whose secret component does not match the one in
+			// the cache.
+			return ctx, fmt.Errorf("API Key %q unauthorized", id)
```

Comment:
Would this case be an unauthorized request? Or the cached key needs to be updated?

Reply:
Nope, this is protecting against a simple attack where a malicious client provides a (possibly invalid) API Key that has the same ID as a valid one in the cache.